### PR TITLE
[new release] lascar (0.6.0)

### PR DESCRIPTION
**Pull request by jserot**
A library for manipulating Labeled Transition Systems in OCaml

- Project page: <a href="github.com/jserot/lascar">github.com/jserot/lascar</a>
- Documentation: <a href="http://jserot.github.io/lascar">http://jserot.github.io/lascar</a>

## 0.6.0 (Sep, 16, 2019)
    * now requires ocaml >= 4.08

## 0.6-alpha (Aug, 18, 2019)
    * removed dependency on camlp4 by rewriting the Fsm_expr, Fsm and ListExt parsers
    * dunified build and install process
    
## 0.5 (Apr, 9, 2018)
	* Library is now packed under Lascar module to prevent name clash
    
## 0.4 (Feb, 5, 2018)
	* OPAMification
	* Home page and documentation moved to separate web site
	* Added INSTALL
	* CHANGELOG now in .md format
	
## 0.3 (Dec, 14, 2017)
	* Added functions [map_state], [map_attr] and [map_label]
	* Added modules [Builtins.{Int,String,Bool}]. Rewrote some examples accordingly
	
## 0.2 (Nov 27, 2107)
    * Added [ListExt.update_assoc]
	
## 0.1 (Nov 6, 2107)
    * First public version

### CODE REVIEW

File: packages/lascar/lascar.0.6.0/opam [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+description: "A library for manipulating Labeled Transition Systems in OCaml"
+maintainer: ["Jocelyn Sérot <jocelyn.serot@uca.fr>"]
+authors: ["Jocelyn Sérot <jocelyn.serot@uca.fr>"]
+license: "MIT"
+homepage: "github.com/jserot/lascar"
+doc: "http://jserot.github.io/lascar"
+bug-reports: "jocelyn.serot@uca.fr"
+depends: [
+  "dune" {>= "1.11"}
+  "ocaml" {>= "4.06"}
```
Comment (kit-ty-kate) [inline packages/lascar/lascar.0.6.0/opam]:
```suggestion
  "ocaml" {>= "4.08"}
```
Seems like this package doesn't work with OCaml 4.06 and 4.07:
```
#=== ERROR while compiling lascar.0.6.0 =======================================#
# context              2.0.5 | linux/x86_64 | ocaml-base-compiler.4.07.1 | pinned(https://github.com/jserot/lascar/releases/download/0.6.0/lascar-0.6.0.tbz)
# path                 ~/.opam/4.07/.opam-switch/build/lascar.0.6.0
# command              ~/.opam/4.07/bin/dune build -p lascar -j 2 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/lascar-15-7ff8c2.env
# output-file          ~/.opam/log/lascar-15-7ff8c2.out
### output ###
#       ocamlc src/lib/.lascar.objs/byte/lascar__Ltsa.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.07/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Ltsa.cmo -c -impl src/lib/ltsa.ml)
# File "src/lib/ltsa.ml", line 207, characters 44-52:
# Error: Unbound module Fun
#       ocamlc src/lib/.lascar.objs/byte/lascar__Lts.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.07/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Lts.cmo -c -impl src/lib/lts.ml)
# File "src/lib/lts.ml", line 239, characters 44-52:
# Error: Unbound module Fun
#     ocamlopt src/lib/.lascar.objs/native/lascar__Ltsa.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.07/bin/ocamlopt.opt -w -40 -g -I src/lib/.lascar.objs/byte -I src/lib/.lascar.objs/native -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/native/lascar__Ltsa.cmx -c -impl src/lib/ltsa.ml)
# File "src/lib/ltsa.ml", line 207, characters 44-52:
# Error: Unbound module Fun
```
```
#=== ERROR while compiling lascar.0.6.0 =======================================#
# context              2.0.5 | linux/x86_64 | ocaml-base-compiler.4.06.1 | pinned(https://github.com/jserot/lascar/releases/download/0.6.0/lascar-0.6.0.tbz)
# path                 ~/.opam/4.06/.opam-switch/build/lascar.0.6.0
# command              ~/.opam/4.06/bin/dune build -p lascar -j 2 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/lascar-15-7ff8c2.env
# output-file          ~/.opam/log/lascar-15-7ff8c2.out
### output ###
#       ocamlc src/utils/.utils.objs/byte/utils__ListExt.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/utils/.utils.objs/byte -intf-suffix .ml -no-alias-deps -open Utils -o src/utils/.utils.objs/byte/utils__ListExt.cmo -c -impl src/utils/listExt.ml)
# File "src/utils/listExt.ml", line 59, characters 16-30:
# Error: Unbound module Stdlib
#       ocamlc src/lib/.lascar.objs/byte/lascar__Builtins.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Builtins.cmo -c -impl src/lib/builtins.ml)
# File "src/lib/builtins.ml", line 16, characters 16-30:
# Error: Unbound module Stdlib
#       ocamlc src/lib/.lascar.objs/byte/lascar__Ltsa.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Ltsa.cmo -c -impl src/lib/ltsa.ml)
# File "src/lib/ltsa.ml", line 207, characters 44-52:
# Error: Unbound module Fun
#       ocamlc src/lib/.lascar.objs/byte/lascar__Fsm.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Fsm.cmo -c -impl src/lib/fsm.ml)
# File "src/lib/fsm.ml", line 89, characters 16-30:
# Error: Unbound module Stdlib
#       ocamlc src/lib/.lascar.objs/byte/lascar__Valuation.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Valuation.cmo -c -impl src/lib/valuation.ml)
# File "src/lib/valuation.ml", line 53, characters 73-87:
# Error: Unbound module Stdlib
#     ocamlopt src/utils/.utils.objs/native/utils__ListExt.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlopt.opt -w -40 -g -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Utils -o src/utils/.utils.objs/native/utils__ListExt.cmx -c -impl src/utils/listExt.ml)
# File "src/utils/listExt.ml", line 59, characters 16-30:
# Error: Unbound module Stdlib
#       ocamlc src/lib/.lascar.objs/byte/lascar__Moore.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Moore.cmo -c -impl src/lib/moore.ml)
# File "src/lib/moore.ml", line 18, characters 22-36:
# Error: Unbound module Stdlib
#       ocamlc src/lib/.lascar.objs/byte/lascar__Lts.{cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlc.opt -w -40 -g -bin-annot -I src/lib/.lascar.objs/byte -I src/utils/.utils.objs/byte -I src/utils/.utils.objs/native -intf-suffix .ml -no-alias-deps -open Lascar -o src/lib/.lascar.objs/byte/lascar__Lts.cmo -c -impl src/lib/lts.ml)
# File "src/lib/lts.ml", line 239, characters 44-52:
# Error: Unbound module Fun
```

**Comment by jserot** [comment]
Yes. The package now requires ocaml >= 4.08, as specified in the `dune-project` file.
I do not understand why the corresponding `opam` file, which has been generated with `dune-release spam pkg` from the `dune-project` file, has `ocaml >= 4.06` in its dependency list. 
Bug in `dune-release` ??

**Comment by kit-ty-kate** [comment]
I guessing that's because you have to regenerate the opam file. I don't know how it's done but I'm pretty sure it's done by dune and not by dune-release.

**Comment by kit-ty-kate** [comment]
Could you also add the appropriate constraints to all the `rfsm` packages? They don't seem to be compatible.

**Comment by jserot** [comment]
The opam file has been regenerated by `dune-release`:
```
$ dune-release opam pkg
[-] Creating opam package description for lascar
[+] Wrote opam package description _build/lascar.0.6.0/opam
```
However, `_build/lascar.0.6.0/opam` contains:
```
opam-version: "2.0"
version: "0.6"
synopsis: "A library for manipulating Labeled Transition Systems in OCaml"
description: "A library for manipulating Labeled Transition Systems in OCaml"
maintainer: ["Jocelyn Sérot <jocelyn.serot@uca.fr>"]
authors: ["Jocelyn Sérot <jocelyn.serot@uca.fr>"]
license: "MIT"
homepage: "github.com/jserot/lascar"
doc: "http://jserot.github.io/lascar"
bug-reports: "jocelyn.serot@uca.fr"
depends: [
  "dune" {>= "1.11"}
  "ocaml" {>= "4.06"}
]
build: [
  ["dune" "subst"] {pinned}
  [
    "dune"
    "build"
    "-p"
    name
    "-j"
    jobs
    "@install"
    "@runtest" {with-test}
    "@doc" {with-doc}
  ]
]
dev-repo: "git+https://github.com/jserot/lascar.git"
url {
  src:
    "https://github.com/jserot/lascar/releases/download/0.6.0/lascar-0.6.0.tbz"
  checksum: [
    "sha256=343190b9c765655e787758db86a82818404dda18b4d2806283b4bde3ced91d26"
    "sha512=2b6534ac438d574228d127b3b4e2cfec17b8b95e95124297c8da351caf34de2bb4ec8a31028a1470ba7f42aa932257293d021140a5a1c2642d1a60a8123d00a0"
  ]
}
```

whereas `./dune-project` says:
```
(lang dune 1.11)
(name lascar)
(source (github jserot/lascar))
(license MIT)
(authors "Jocelyn Sérot <jocelyn.serot@uca.fr>")
(maintainers "Jocelyn Sérot <jocelyn.serot@uca.fr>")
(bug_reports "jocelyn.serot@uca.fr")
(homepage "http://github.com/jserot/lascar")
(documentation "http://jserot.github.io/lascar")

(package
  (name lascar)
  (synopsis "A library for manipulating Labeled Transition Systems in OCaml")
  (description "A library for manipulating Labeled Transition Systems in OCaml")
  (depends
    (dune (>= 1.11))
    (ocaml (>= 4.08))))

(generate_opam_files true)
```

**Comment by jserot** [comment]
> Could you also add the appropriate constraints to all the rfsm packages? They don't seem to be compatible.

It is planned. But, since `rfsm` depends on `lascar` i need to have `lascar` correctly packaged before.

**Comment by kit-ty-kate** [comment]
that's what I'm saying, dune-release doesn't generate the opam file, it only adds the corresponding url field to the file but it doesn't modify anything else. Dune itself is responsible for the generation

**Comment by kit-ty-kate** [comment]
as for rfsm I'd rather see all the previous versions fixed before merging this PR as it could break people using it in between the merge of the two prs

**Comment by jserot** [comment]
Ah ok. I understand now : `dune-release` has used the old `opam` file. I should have instructed `dune` to regenerate it before invoking it.. What's the correct way to do this ?

**Comment by kit-ty-kate** [comment]
I have no idea. `dune build` doesn't work?

**Comment by kit-ty-kate** [comment]
or try `dune build @all` if it doesn't work

**Comment by jserot** [comment]
> dune build doesn't work?

No. 

`dune build @install` does !

**Comment by jserot** [comment]
> as for rfsm I'd rather see all the previous versions fixed before merging this PR as it could break people using it in between the merge of the two prs

What do you mean by _fixing all the previous versions_ ? 
Not the previous versions of  `lascar` - since they were not 4.08 ready ?

**Comment by kit-ty-kate** [comment]
> > as for rfsm I'd rather see all the previous versions fixed before merging this PR as it could break people using it in between the merge of the two prs
> 
> What do you mean by _fixing all the previous versions_ ?
> Not the previous versions of `lascar` - since they were not 4.08 ready ?

I meant all the versions of `rfsm`. Currently with this PR they all fail with something like:
```
# File "src/lib/static.ml", line 227, characters 26-33:
# 227 |        (ListExt.to_string Misc.id "," rdrs)
#                                 ^^^^^^^
# Error: Unbound value Misc.id
```

**Comment by jserot** [comment]
Do you mean i should open a PR for the 4.08-ready version of  `rfsm` ? 
If yes, there's a circular dependency problem since this version will have `lascar.0.6.0` in its dependency list !

**Comment by kit-ty-kate** [comment]
> Do you mean i should open a PR for the 4.08-ready version of `rfsm` ?

No, `rfsm.1.0`, `1.4.2`, `1.5` and `1.6-alpha-3` will break with this PR. I meant these versions need to have their constraints on `lascar` fixed. If you really don't have time to do it, I'll do it.

**Comment by jserot** [comment]
You mean, for ex, that `rfsm-1.5` should have `lascar < 0.6` and `ocaml < 4.08` in its constraints so that this version of `lascar` will not try to build it ?
It's not time i'm missing for handling this, just skill, i'm afraid ;-)

**Comment by kit-ty-kate** [comment]
> You mean, for ex, that `rfsm-1.5` should have `lascar < 0.6` and `ocaml < 4.08` in its constraints so that this version of `lascar` will not try to build it ?

Yes, only `lascar < 0.6` should suffice though, they work perfectly with OCaml 4.08 at the time.

**Comment by jserot** [comment]
> Yes, only lascar < 0.6 should suffice though, they work perfectly with OCaml 4.08 at the time.

??
 They shouldn't since they use the `Pervasive` module which is deprecated since OCaml 4.08 IIUC

**Comment by kit-ty-kate** [comment]
> > Yes, only lascar < 0.6 should suffice though, they work perfectly with OCaml 4.08 at the time.
> 
> ??
> They shouldn't since they use the `Pervasive` module which is deprecated since OCaml 4.08 IIUC

Deprecated yes, but it still compiles and works, nothing is different for its users. It will only require a constraint when those deprecation warnings will become proper errors in a future version of OCaml (probably in like 3 - 4 years, or something)

**Comment by jserot** [comment]
Dune seems to view it as an error. E.g.:

```
$ cat foo.ml       
let _ = Pervasives.compare 1 2
$ dune build ./foo.exe
File "foo.ml", line 1, characters 8-26:
1 | let _ = Pervasives.compare 1 2
            ^^^^^^^^^^^^^^^^^^
Error (alert deprecated): module Stdlib.Pervasives
Use Stdlib instead.
If you need to stay compatible with OCaml < 4.07, you can use the 
stdlib-shims library: https://github.com/ocaml/stdlib-shims
```

**Comment by jserot** [comment]
Btw, the next version of `rfsm` is now ready for PR submission. So maybe it's not necessary to fix the old versions ?

**Comment by mseri** [comment]
The bound on the old versions is needed independently of the new release. Recent versions of dune will fail only in dev mode, release builds will only show the warnings. For the rest, @kit-ty-kate is there any other action needed before merge?

**Comment by kit-ty-kate** [comment]
> Dune seems to view it as an error. E.g.:
> 
> ```
> $ cat foo.ml       
> let _ = Pervasives.compare 1 2
> $ dune build ./foo.exe
> File "foo.ml", line 1, characters 8-26:
> 1 | let _ = Pervasives.compare 1 2
>             ^^^^^^^^^^^^^^^^^^
> Error (alert deprecated): module Stdlib.Pervasives
> Use Stdlib instead.
> If you need to stay compatible with OCaml < 4.07, you can use the 
> stdlib-shims library: https://github.com/ocaml/stdlib-shims
> ```

This is the behaviour for development purposes (i.e. when you launch `dune build` by default), for opam packages, `-p` is used which also turns `--profile=release` which does not turn any warnings as errors. You can see here the current status of your packages: http://check.ocamllabs.io/?maintainers=Jocelyn+S%C3%A9rot

**Comment by kit-ty-kate** [comment]
> For the rest, @kit-ty-kate is there any other action needed before merge?

Yes there is still a change to  be made here which hasn't been pushed: https://github.com/ocaml/opam-repository/pull/14827#issuecomment-531756415 (the generated opam file is not up to date

**Comment by kit-ty-kate** [comment]
> Btw, the next version of `rfsm` is now ready for PR submission. So maybe it's not necessary to fix the old versions ?

In your next submission be careful about the version number. The next release should be `rfsm.1.6.0` and not `rfsm.1.6` otherwise `rfsm.1.6-alpha-3` will be given priority over `1.6` (see The "version ordering` section in the manual: https://opam.ocaml.org/doc/Manual.html#Package-Formulas)

**Comment by jserot** [comment]
> In your next submission be careful about the version number. he next release should be rfsm.1.6.0 and not rfsm.1.6

Ok. Shall i start the PR now or do i have to wait for this one to be completed ?

**Comment by jserot** [comment]
> This one should preferably be merged before. Could you push/publish the change discussed above in #14827 (comment)?

Done

**Comment by kit-ty-kate** [comment]
Thanks!

**Comment by jserot** [comment]
Thanks to both of you !

**Pull request merged**